c10/core/impl/SizesAndStrides.h

```diff
@@ -73,46 +194,120 @@ class C10_API SizesAndStrides {
+    const auto oldSize = size();
+    if (newSize == oldSize) {
+      return;
+    }
+    if (C10_LIKELY(newSize <= MAX_INLINE_SIZE)) {
+      if (C10_LIKELY(isInline())) {
+        if (oldSize < newSize) {
+          const auto bytesToZero = (newSize - oldSize) * sizeof(inlineStorage_[0]);
+          memset(&inlineStorage_[oldSize], 0, bytesToZero);
+          memset(&inlineStorage_[MAX_INLINE_SIZE + oldSize], 0, bytesToZero);
+        }
+      } else {
+        memcpy(&inlineStorage_[0], outOfLineStorage()->data(), MAX_INLINE_SIZE * sizeof(inlineStorage_[0]));
+        memcpy(
+            &inlineStorage_[MAX_INLINE_SIZE],
+            outOfLineStorage()->data() + outOfLineStorage()->size() / 2,
+            MAX_INLINE_SIZE * sizeof(inlineStorage_[0]));
+        delete outOfLineStorage();
+      }
+      setInlineSize(newSize);
+    } else {
+      if (isInline()) {
+        taggedStorageOrSize_ = reinterpret_cast<int64_t>(new std::vector<int64_t>(newSize * 2));
+        const auto bytesToCopy = oldSize * sizeof(inlineStorage_[0]);
+        memcpy(outOfLineStorage()->data(), &inlineStorage_[0], bytesToCopy);
+        memcpy(outOfLineStorage()->data() + newSize, &inlineStorage_[MAX_INLINE_SIZE], bytesToCopy);
+      } else {
+        const bool isGrowing = oldSize < newSize;
+        if (isGrowing) {
+          // Resize before shifting so that we have room.
+          outOfLineStorage()->resize(newSize * 2);
+        }
+        // Shift the old strides to their new starting point. Note
+        // that this does not occur in the inline path above because
+        // the stride starting point is not moving.
+        memmove(
+            outOfLineStorage()->data() + newSize,
+            outOfLineStorage()->data() + oldSize,
+            std::min(oldSize, newSize) * sizeof(inlineStorage_[0]));
+        if (!isGrowing) {
+          // Resize after shifting so that we don't lose data.
+          outOfLineStorage()->resize(newSize * 2);
+        } else {
+          // Zero the end of the sizes portion.
+          memset(outOfLineStorage()->data() + oldSize, 0, (newSize - oldSize) * sizeof(inlineStorage_[0]));
+        }
+      }
+    }
   }
 
  private:
-  SmallVector<int64_t,5> sizes_;
-  SmallVector<int64_t,5> strides_;
+  bool isInline() const {
+    return (taggedStorageOrSize_ & 1) == 1;
+  }
+
+  std::vector<int64_t>* outOfLineStorage() const {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!isInline());
+    return reinterpret_cast<std::vector<int64_t>*>(taggedStorageOrSize_);
+  }
+
+  void resetToInlineStorage() {
+    setInlineSize(1);
+  }
+
+  void setInlineSize(size_t sz) {
+    taggedStorageOrSize_ = (sz << 1) | 1;
+  }
+
+  void copyFromInline(const SizesAndStrides& rhs) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(rhs.isInline());
+    taggedStorageOrSize_ = rhs.taggedStorageOrSize_;
+    memcpy(inlineStorage_, rhs.inlineStorage_, sizeof(inlineStorage_));
+  }
+
+  static constexpr int MAX_INLINE_SIZE = 5;
+  int64_t taggedStorageOrSize_;
```

Comment:
nit: wouldn't `intptr_t` be better here?

Reply:
NB: when it's a pointer, this is `std::vector<int64_t>* taggedStorageOrSize_`, which means there's actually a *double indirection* when you overflow the inline fields. I can see using a `std::vector` makes implementation a lot more convenient, so maybe it's OK for now, but while the patient is open it might be best to do "the best we can" implementation-wise.

Reply:
> while the patient is open it might be best to do "the best we can" 

This runs counter to the philosophy that diffs should be small and focused. If performance in the out-of-line case is important for some reason, we can spend a second diff focusing on that. (But if it's important, that invalidates the assumption underlying this implementation -- we are intentionally optimizing rank <= 5 at the expense of larger ranks!)

Reply:
That said, I'll switch representations as I mentioned in the other subthread because it becomes easy to get rid of the vector.

Reply:
> This runs counter to the philosophy that diffs should be small and focused. If performance in the out-of-line case is important for some reason, we can spend a second diff focusing on that. 

Acknowledged. I'm mostly looking for opportunities along the lines of "it's the same net amount of work, but is ever so slightly more optimized." If it's a lot more work definitely OK for you to  push back like that (though, I'm glad you liked my other suggestion!)